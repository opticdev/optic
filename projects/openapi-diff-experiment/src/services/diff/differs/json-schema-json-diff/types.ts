import { BodyLocation, OpenAPIV3 } from '@useoptic/openapi-utilities';
import Ajv, { ErrorObject } from 'ajv';
import { ShapeDiffTypes } from '../../types';
import { JsonSchemaPatch } from './plugins/plugin-types';
import { JsonPatcher } from '../../../patch/incremental-json-patch/json-patcher';

export interface JsonSchemaJsonDiffer {
  ajvToDiff: (
    schema: OpenAPIV3.SchemaObject,
    schemaPath: string,
    diff: ErrorObject,
    example: any,
    location: BodyLocation
  ) => ShapeDiffTypes;
  diffToPatch: (
    diff: ShapeDiffTypes,
    patcher: JsonPatcher<OpenAPIV3.Document>
  ) => JsonSchemaPatch[];
  compare: (
    schema: OpenAPIV3.SchemaObject,
    to: any,
    location: BodyLocation,
    schemaPath: string,
    options: {
      collapseToFirstInstanceOfArrayDiffs: boolean;
    }
  ) => ShapeDiffTypes[];
  supportedKeywords: string[];
  opticAjv: Ajv;
  stripKeywordsNotSupported: (diffs: ErrorObject[]) => ErrorObject[];
}
