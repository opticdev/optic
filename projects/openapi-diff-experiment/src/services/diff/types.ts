/*

  Interface diffing traffic
   - produces a diff from the
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
import { ApiTraffic } from '../traffic/types';
import { OpenAPIDiffingQuestions } from '../read/types';
import {
  ConceptualLocation,
  FieldLocation,
  OpenAPIV3,
} from '@useoptic/openapi-utilities';
import { JsonSchemaKnownKeyword } from './differs/json-schema-json-diff/plugins/plugin-types';
import { JsonSchemaJsonDiffer } from './differs/json-schema-json-diff/types';
import { JsonPath } from '@useoptic/openapi-io';

export interface IDiffService {
  compare(traffic: ApiTraffic): Promise<{ diffs: IDiff[]; errors: string[] }>;
  jsonSchemaDiffer: JsonSchemaJsonDiffer;
}

export type DiffServiceFactoryFunction = (
  spec: OpenAPIDiffingQuestions
) => IDiffService;

export enum DiffType {
  UnmatchedPath = 'UnmatchedPath',
  UnmatchedMethod = 'UnmatchedMethod',
  UnmatchedResponse = 'UnmatchedResponse',
  BodyAdditionalProperty = 'BodyAdditionalProperty',
  BodyMissingRequiredProperty = 'BodyMissingRequiredProperty',
  BodyUnmatchedType = 'BodyUnmatchedType',
  QueryAdditionalParameter = 'QueryAdditionalParameter',
}

export interface IDiff {
  type: DiffType;
}

export interface UnmatchedPath extends IDiff {
  type: DiffType.UnmatchedPath;
  path: string;
  method: OpenAPIV3.HttpMethods;
  closestMatch: string;
}

export interface UnmatchedMethod extends IDiff {
  type: DiffType.UnmatchedMethod;
  path: string;
  method: OpenAPIV3.HttpMethods;
}

export interface UnmatchedResponse extends IDiff {
  type: DiffType.UnmatchedResponse;
  path: string;
  method: OpenAPIV3.HttpMethods;
  statusCode: string;
}
////////////////////////////////////////////////////////////

export interface QueryAdditionalParameter extends IDiff {
  type: DiffType.QueryAdditionalParameter;
  path: string;
  name: string;
  method: OpenAPIV3.HttpMethods;
  example: any;
}

////////////////////////////////////////////////////////////

export interface BodyAdditionalProperty extends IDiff {
  type: DiffType.BodyAdditionalProperty;
  propertyExamplePath: JsonPath;
  parentObjectPath: JsonPath;
  key: string;
  example: any;
  propertyPath: JsonPath;
  keyword: JsonSchemaKnownKeyword.additionalProperties;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export interface BodyMissingRequiredProperty extends IDiff {
  type: DiffType.BodyMissingRequiredProperty;
  keyword: JsonSchemaKnownKeyword.required;
  parentObjectPath: JsonPath;
  propertyPath: JsonPath;
  key: string;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export interface BodyPropertyUnmatchedType extends IDiff {
  type: DiffType.BodyUnmatchedType;
  keyword: JsonSchemaKnownKeyword.type | JsonSchemaKnownKeyword.oneOf;
  propertyPath: JsonPath;
  key: string;
  example: any;
  schemaPath: JsonPath;
  instancePath: JsonPath;
  location: FieldLocation;
}

export type ShapeDiffTypes =
  | BodyMissingRequiredProperty
  | BodyAdditionalProperty
  | BodyPropertyUnmatchedType;

type DiffKinds =
  | UnmatchedMethod
  | UnmatchedPath
  | UnmatchedResponse
  | QueryAdditionalParameter
  | ShapeDiffTypes;

/*
  Diff Result Types
 */

export interface DiffResult {
  isMatch: boolean;
}
export interface MatchedWithContext<Context> extends DiffResult {
  isMatch: true;
  context: Context;
}
export interface Matched extends DiffResult {
  isMatch: true;
}

export interface Diff extends DiffResult {
  isMatch: false;
  diffs: DiffKinds[];
}

export interface DiffError extends DiffResult {
  isMatch: false;
  error: string;
}

export interface EitherDiffResult<Context = undefined> {
  context: Context;
  error: string;
  diffs: DiffKinds[];
  isMatch: boolean;
  isDiff: boolean;
}

export function EitherDiffResult<Context>(
  i: DiffResult
): EitherDiffResult<Context> {
  return {
    isDiff: !i.isMatch,
    isMatch: i.isMatch,
    context: i.hasOwnProperty('context')
      ? (i as MatchedWithContext<Context>).context
      : undefined,
    error: i.hasOwnProperty('error') ? (i as DiffError).error : undefined,
    diffs: i.hasOwnProperty('diffs') ? (i as Diff).diffs : [],
  };
}

export const DiffResult = {
  match: () => {
    return EitherDiffResult<undefined>({ isMatch: true } as Matched);
  },
  matchWithContext: <Context>(context: Context) => {
    return EitherDiffResult<Context>({
      isMatch: true,
      context,
    } as MatchedWithContext<Context>);
  },
  diff: (diffs: DiffKinds[]) => {
    return EitherDiffResult<undefined>({
      isMatch: false,
      diffs,
    } as Diff);
  },
  error: (error: string) => {
    return EitherDiffResult<undefined>({ isMatch: false, error } as DiffError);
  },
};

export function isSchemaDiff(i: IDiff): boolean {
  return (
    i.type === DiffType.BodyAdditionalProperty ||
    i.type === DiffType.BodyUnmatchedType ||
    i.type === DiffType.BodyMissingRequiredProperty
  );
}
