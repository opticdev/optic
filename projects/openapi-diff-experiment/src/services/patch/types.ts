import { OpenAPIV3 } from '@useoptic/openapi-utilities';
import { ISpecReader } from '../read/types';
import {
  IPatchGroup,
  JsonPatcher,
  PatchesToSave,
} from './incremental-json-patch/json-patcher';
import { ApiTraffic } from '../traffic/types';
import { ShapeDiffTypes } from '../diff/types';
import { JsonSchemaPatch } from '../diff/differs/json-schema-json-diff/plugins/plugin-types';

/*
  Patching a specification has two steps
  - mutating the JSON-like object, by exposing domain-relevant helpers
  - updating the JSON-like objects across the file system to match w/ what we have in-memory via a reconcile pattern. Json-like * sourcemap
    essentially redraw the JSON across the file system with our changes
 */

/*
  During WIP Phase, we will not be reconciling the file system -- instead we will simple, be flushing via stringify methods.
  It's destructive, but not the part to optimize first
 */

export interface IPatchOpenAPI {
  reset: () => void;
  listPatches: () => IPatchGroup[];
  save: (options: { dryRun: boolean }) => Promise<IFilePatch>;
  fork: () => Promise<IPatchOpenAPI>;
  forkedPatcher(): JsonPatcher<OpenAPIV3.Document>;
  init: {
    operation: (
      pathPattern: string,
      method: OpenAPIV3.HttpMethods,
      example: ApiTraffic
    ) => void;
    queryParameter: (
      method: OpenAPIV3.HttpMethods,
      path: string,
      name: string,
      exampleValue: string | boolean | number
    ) => void;
    response: (
      pathPattern: string,
      method: OpenAPIV3.HttpMethods,
      example: ApiTraffic
    ) => void;
    responseBody: (
      pathPattern: string,
      method: OpenAPIV3.HttpMethods,
      statusCodeMatcher: string,
      contentType: string,
      example: ApiTraffic
    ) => void;
  };

  patch: {
    property: (jsonSchemaPatch: JsonSchemaPatch) => void;
  };
}

export interface IPatchOpenAPIReconciler {
  patchesToFileMutations(
    patchesToSave: PatchesToSave<OpenAPIV3.Document>
  ): Promise<IFilePatch>;
}

export type IPatchOpenAPIReconcilerFactory = (
  reader: ISpecReader
) => IPatchOpenAPIReconciler;

export type IFilePatch = {
  files: { path: string; previousContents: string; newContents: string }[];
};
