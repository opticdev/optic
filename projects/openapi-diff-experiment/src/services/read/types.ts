/*

  OpenAPI projections for diffs
   - assemble all the information required to diff
   - retrain pointers to ast nodes

   */
import {
  BodyLocation,
  OpenAPIV3,
  QueryParameterLocation,
} from '@useoptic/openapi-utilities';
import { IFilePatch } from '../patch/types';
import { JsonSchemaSourcemap } from '@useoptic/openapi-io';

export interface OpenAPIDiffingQuestions {
  paths(): string[];
  operations(): {
    path: string;
    method: OpenAPIV3.HttpMethods;
    jsonPath: string;
  }[];
  responsesForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): ResponseMatchType[];

  requestBodiesForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): RequestBodyMatchType[];

  queryParametersForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): QueryParameterType[];
}

export interface ISpecReader {
  questions(): Promise<OpenAPIDiffingQuestions>;
  sourcemap(): Promise<JsonSchemaSourcemap>;
  flattenedSpecification(): Promise<OpenAPIV3.Document>;
  didLoad(): Promise<DidLoadStatus>;
  save(patch: IFilePatch): Promise<void>;
  describeLocation(): string;
  rootFile(): string;
  reload(): Promise<void>;
  mode: 'simulated' | 'filesystem';
}

export type FilePathsWithChanges = string[];
export type DidLoadStatus =
  | { success: true; durationMillis: number }
  | { success: false; error: string; durationMillis: number };

export const openApiDiffingQuestionsTestingStub: OpenAPIDiffingQuestions = {
  operations() {
    return [];
  },
  paths(): string[] {
    return [];
  },
  requestBodiesForOperation(method: OpenAPIV3.HttpMethods, path: string) {
    return [];
  },
  responsesForOperation(method: OpenAPIV3.HttpMethods, path: string) {
    return [];
  },
  queryParametersForOperation(method: OpenAPIV3.HttpMethods, path: string) {
    return [];
  },
};

// Result Types
export type ResponseMatchType = {
  statusCodeMatcher: string;
  contentTypes: {
    contentType: string;
    schema?: OpenAPIV3.SchemaObject;
    location: BodyLocation;
    jsonPath: string;
  }[];
};

export type RequestBodyMatchType = {
  contentType: string;
  schema?: OpenAPIV3.SchemaObject;
  location: BodyLocation;
  jsonPath: string;
};

export type QueryParameterType = {
  name: string;
  schema?: OpenAPIV3.ReferenceObject | OpenAPIV3.SchemaObject;
  location: QueryParameterLocation;
  jsonPath: string;
  required: boolean;
};
