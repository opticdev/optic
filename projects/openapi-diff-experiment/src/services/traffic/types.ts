import events from 'events';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
/*

  Interface for traffic, expose what the differ cares about via interfaces
   - different formats can have different implementations
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
export interface ApiTraffic {
  method: OpenAPIV3.HttpMethods;
  path: string;
  queryString: string;
  response: {
    statusCode: string;
    body: {
      contentType?: string;
      jsonBodyString?: string;
    };
  };
  requestBody?: {
    contentType?: string;
    jsonBodyString?: string;
  };
}

/*

  Interface for traffic, should get traffic (somehow), and emit each interaction into our queue
   - each implementation should be self contained starting / parsing / stopping traffic collection
   - known types
     - file
     - in-memory (for debugging)
     - proxy
     - postman
     - sniff
     - chrome

   */

export interface ITrafficSource {
  start(): Promise<void>;
  stop(): Promise<void>;
  emitTraffic(traffic: ApiTraffic): Promise<void>;
  on(event: 'traffic', listener: (example: ApiTraffic) => void): this;
}

export class TrafficSource
  extends events.EventEmitter
  implements ITrafficSource
{
  emitTraffic(traffic: ApiTraffic): Promise<void> {
    this.emit('traffic', traffic);
    return Promise.resolve();
  }

  start(): Promise<void> {
    return Promise.resolve(undefined);
  }

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}
