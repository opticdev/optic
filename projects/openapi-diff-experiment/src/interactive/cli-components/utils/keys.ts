export const keys = {
  leftArrow: '←',
  rightArrow: '→',
};
