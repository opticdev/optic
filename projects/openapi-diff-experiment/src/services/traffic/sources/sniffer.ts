import { ChildProcessWithoutNullStreams, spawn } from 'child_process';

import { TrafficSource } from '../types';
import { OpticHttpInteraction } from '../traffic/optic-http-interaction';

export type SnifferConfig = {
  interface: string;
  port: number;
};

export class SnifferSource extends TrafficSource {
  private capture: ChildProcessWithoutNullStreams;

  constructor(private config: SnifferConfig) {
    super();
  }

  async start(): Promise<void> {
    // if (captureToolNotFound...) {}
    // throw BOOM
    /*
      ray's magic

     */
    // this.emitTraffic()

    // TODO: Actually check that we can run the command
    const capture_binary_path = process.env['CAPTURE_PATH'] || 'optic_capture';
    this.capture = spawn('sudo', [
      capture_binary_path,
      'sniff',
      `:${this.config.port}`,
      '--interface',
      `${this.config.interface}`,
      'output',
      'stdout',
    ]);

    this.capture.on('error', (err: any) => {
      throw `Cannot start optic_capture ${err}`;
    });

    this.capture.on('error', (code: number, signal: any) => {
      if (code != 0) {
        throw `Capture exited with code ${code}`;
      }
    });

    var startupOk = new Promise<void>((resolve) => {
      this.capture.stderr.on('data', (data: any) => {
        //console.log("%s", data);
        resolve();
      });

      this.capture.stdout.on('data', async (data: string) => {
        resolve();
        //console.log("%s", data);
        await this.emitTraffic(new OpticHttpInteraction(JSON.parse(data)));
      });
    });
    await startupOk;
  }

  async stop(): Promise<void> {
    await this.capture.kill('SIGTERM');
  }
}
