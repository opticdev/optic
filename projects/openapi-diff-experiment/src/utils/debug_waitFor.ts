export async function waitFor(millis: number) {
  return await new Promise((resolve) => {
    setTimeout(resolve, millis);
  });
}

export function copyObject<G>(before: any): G {
  return JSON.parse(JSON.stringify(before));
}
