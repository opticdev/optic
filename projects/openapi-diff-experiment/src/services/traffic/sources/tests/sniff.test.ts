import os from 'os';
import { AddressInfo } from 'net';
import http from 'http';
import fetch from 'node-fetch';

import express from "express";

import { SnifferSource } from '../sniffer';
import { ApiTraffic } from '../../types';
import { waitFor } from '../../../../utils/debug_waitFor';

const rootOnlyIt = os.userInfo().uid == 0 ? it : it.skip;

describe('capture sources', () => {
  rootOnlyIt('can parse captured traffic', async () => {
    let server = startServer();
    let addr: AddressInfo | string = server.address() as AddressInfo;
    console.log(`server addr ${JSON.stringify(server.address())}`);

    const source = new SnifferSource({interface: 'lo0', port: addr.port});


    const examples: ApiTraffic[] = [];
    source.on('traffic', (example) => {
      //console.log(example);
      examples.push(example);
    });

    await source.start();

    for (var i = 0; i < 4; i++) {
      let url = makeTestURL(server, i)
      await fetch(url);
    }

    await waitFor(300);
    await source.stop();
    server.close();

    expect(JSON.stringify(examples)).toMatchSnapshot();
  });
});

function startServer(): http.Server {
  const app = express();
  const port = 0;
  
  app.get('/get', (req: any, res: any) => {
    res.send(JSON.stringify(req.query));
  });
  
  return app.listen(port);
}

function makeTestURL(server: http.Server, id: number): string {
    let addr: AddressInfo | string = server.address() as AddressInfo;
    if (addr.family == "IPv6") {
      return `http://[${addr.address}]:${addr.port}/get?i=${id}`;
    } else {
      return `http://${addr.address}:${addr.port}/get?i=${id}`;
    }
}
