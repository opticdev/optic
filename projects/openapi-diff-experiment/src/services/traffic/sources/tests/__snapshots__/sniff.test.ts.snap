// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`capture sources can parse captured traffic 1`] = `"[{\\"_method\\":\\"get\\",\\"_path\\":\\"/get\\",\\"_response\\":{\\"statusCode\\":200,\\"body\\":{\\"contentType\\":\\"application/json\\"}}},{\\"_method\\":\\"get\\",\\"_path\\":\\"/get\\",\\"_response\\":{\\"statusCode\\":200,\\"body\\":{\\"contentType\\":\\"application/json\\"}}},{\\"_method\\":\\"get\\",\\"_path\\":\\"/get\\",\\"_response\\":{\\"statusCode\\":200,\\"body\\":{\\"contentType\\":\\"application/json\\"}}},{\\"_method\\":\\"get\\",\\"_path\\":\\"/get\\",\\"_response\\":{\\"statusCode\\":200,\\"body\\":{\\"contentType\\":\\"application/json\\"}}}]"`;
